public class org.kobjects.util.ChainedRuntimeException extends java.lang.RuntimeException
{
    java.lang.Exception chain;

    void <init>()
    {
        org.kobjects.util.ChainedRuntimeException r0;

        r0 := @this: org.kobjects.util.ChainedRuntimeException;

        specialinvoke r0.<java.lang.RuntimeException: void <init>()>();

        return;
    }

    void <init>(java.lang.Exception, java.lang.String)
    {
        org.kobjects.util.ChainedRuntimeException r0;
        java.lang.StringBuilder $r3;
        java.lang.String $r2;
        java.lang.Exception $r1;

        r0 := @this: org.kobjects.util.ChainedRuntimeException;

        $r1 := @parameter0: java.lang.Exception;

        $r2 := @parameter1: java.lang.String;

        $r3 = new java.lang.StringBuilder;

        specialinvoke $r3.<java.lang.StringBuilder: void <init>()>();

        if $r2 != null goto label1;

        $r2 = "rethrown";

     label1:
        $r3 = virtualinvoke $r3.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.String)>($r2);

        $r3 = virtualinvoke $r3.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.String)>(": ");

        $r2 = virtualinvoke $r1.<java.lang.Exception: java.lang.String toString()>();

        $r3 = virtualinvoke $r3.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.String)>($r2);

        $r2 = virtualinvoke $r3.<java.lang.StringBuilder: java.lang.String toString()>();

        specialinvoke r0.<java.lang.RuntimeException: void <init>(java.lang.String)>($r2);

        r0.<org.kobjects.util.ChainedRuntimeException: java.lang.Exception chain> = $r1;

        return;
    }

    public static org.kobjects.util.ChainedRuntimeException create(java.lang.Exception, java.lang.String)
    {
        org.kobjects.util.ChainedRuntimeException $r4;
        java.lang.Class $r2;
        java.lang.Object $r3;
        java.lang.String $r1;
        java.lang.Exception $r0, $r5;

        $r0 := @parameter0: java.lang.Exception;

        $r1 := @parameter1: java.lang.String;

     label1:
        $r2 = staticinvoke <java.lang.Class: java.lang.Class forName(java.lang.String)>("org.kobjects.util.ChainedRuntimeExceptionSE");

        $r3 = virtualinvoke $r2.<java.lang.Class: java.lang.Object newInstance()>();

        $r4 = (org.kobjects.util.ChainedRuntimeException) $r3;

        $r4 = virtualinvoke $r4.<org.kobjects.util.ChainedRuntimeException: org.kobjects.util.ChainedRuntimeException _create(java.lang.Exception,java.lang.String)>($r0, $r1);

     label2:
        return $r4;

     label3:
        $r5 := @caughtexception;

        $r4 = new org.kobjects.util.ChainedRuntimeException;

        specialinvoke $r4.<org.kobjects.util.ChainedRuntimeException: void <init>(java.lang.Exception,java.lang.String)>($r0, $r1);

        return $r4;

        catch java.lang.Exception from label1 to label2 with label3;
    }

    org.kobjects.util.ChainedRuntimeException _create(java.lang.Exception, java.lang.String)
    {
        org.kobjects.util.ChainedRuntimeException r0;
        java.lang.String $r2;
        java.lang.RuntimeException $r3;
        java.lang.Exception $r1;

        r0 := @this: org.kobjects.util.ChainedRuntimeException;

        $r1 := @parameter0: java.lang.Exception;

        $r2 := @parameter1: java.lang.String;

        $r3 = new java.lang.RuntimeException;

        specialinvoke $r3.<java.lang.RuntimeException: void <init>(java.lang.String)>("ERR!");

        throw $r3;
    }

    public java.lang.Exception getChained()
    {
        org.kobjects.util.ChainedRuntimeException r0;
        java.lang.Exception r1;

        r0 := @this: org.kobjects.util.ChainedRuntimeException;

        r1 = r0.<org.kobjects.util.ChainedRuntimeException: java.lang.Exception chain>;

        return r1;
    }

    public void printStackTrace()
    {
        org.kobjects.util.ChainedRuntimeException r0;
        java.lang.Exception $r1;

        r0 := @this: org.kobjects.util.ChainedRuntimeException;

        specialinvoke r0.<java.lang.RuntimeException: void printStackTrace()>();

        $r1 = r0.<org.kobjects.util.ChainedRuntimeException: java.lang.Exception chain>;

        if $r1 == null goto label1;

        $r1 = r0.<org.kobjects.util.ChainedRuntimeException: java.lang.Exception chain>;

        virtualinvoke $r1.<java.lang.Exception: void printStackTrace()>();

     label1:
        return;
    }
}
